{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Example workflow for running varDFE to infer the DFE in a single species\n",
    "\n",
    "Last updated: 2024-10-29"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import varDFE\n",
    "import dadi"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## step0: load and visualize the data\n",
    "\n",
    "This example demonstrates the workflow using a site frequency spectrum (SFS) derived from The 1000 Genomes Project data. \n",
    "For details on how this SFS was generated, refer to [Zhen et al. 2021](https://genome.cshlp.org/content/31/1/110).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read and fold the spectra\n",
    "synfs = dadi.Spectrum.from_file('SYN.sfs') # synonymous\n",
    "misfs = dadi.Spectrum.from_file('MIS.sfs') # nonsynonymous/missense"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate general information\n",
    "print(synfs.S())\n",
    "print(misfs.S())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# adapt a plotting function from varDFE\n",
    "# here the model is missense sfs and data is synonymous sfs (you can change the naming later)\n",
    "%matplotlib inline\n",
    "from varDFE.Misc.Plotting import ggplot_dadi_1d\n",
    "ggplot_dadi_1d(outprefix = 'MIS_SYN', model = misfs, fs = synfs, yvar = 'count', returnplot=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make a percentage plot\n",
    "%matplotlib inline\n",
    "varDFE.Misc.Plotting.ggplot_dadi_1d(outprefix = 'MIS_SYN', model = misfs, fs = synfs, yvar = 'percent', returnplot=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## step1: `Demog1D_sizechangeFIM` module for demographic inference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "mkdir -p ./output/logs/\n",
    "WORKSCRIPT='../workflow/Demography/Demog1D_sizechangeFIM.py'\n",
    "\n",
    "# show help\n",
    "python $WORKSCRIPT -h\n",
    "\n",
    "# run demographic inference\n",
    "python $WORKSCRIPT \\\n",
    "--pop 'example' --mu '2.50E-08' --Lcds '19089129' --NS_S_scaling '2.31' --Nrun 5 \\\n",
    "'SYN.sfs' 'two_epoch' './output/demography/two_epoch' &> './output/logs/demography.log'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Command line explanation\n",
    "\n",
    "Positional arguments:\n",
    "|Position | Value | Explanation |\n",
    "|---------|-------|-------------|\n",
    "|1 | 'SYN.sfs' | Path to folded synonymous SFS file |\n",
    "|2 | 'two_epoch' | Demographic model to use |\n",
    "|3 | './output/demography/two_epoch' | Output directory path |\n",
    "\n",
    "Options:\n",
    "|Argument | Value | Explanation |\n",
    "|---------|-------|-------------|\n",
    "|--pop | 'example' | Population identifier |\n",
    "|--mu | '2.50E-08' | Exon mutation rate (mutations/bp/generation) |\n",
    "|--Lcds | '19089129' | Number of called CDS sites used in SFS |\n",
    "|--NS_S_scaling | '2.31' | Scaling factor for Lsyn and Lnonsyn length |\n",
    "|--Nrun | 5 | Number of replicated runs |\n",
    "\n",
    "Additional:\n",
    "- Redirects both stdout and stderr to './output/logs/demography.log'\n",
    "\n",
    "### Log outputs\n",
    "\n",
    "This script gives (hopefully) most of the information you need for demographic inference and comparisons:\n",
    "\n",
    "In the output log, it provides:\n",
    "\n",
    "1. Input parameters parsed\n",
    "2. Within each replicated inference, every iteration's log-likelihood and demographic parameter values\n",
    "3. Top three replicates with the highest log-likelihoods\n",
    "4. Convergence of the best inference with the highest log-likelihoods\n",
    "5. Fisher's Information Matrix (FIM) based standard deviation estimates\n",
    "\n",
    "### File outputs\n",
    "\n",
    "In the output folders, the files are organized as following:\n",
    "\n",
    "```\n",
    ".\n",
    "├── demography\n",
    "│   └── two_epoch\n",
    "│       ├── bestrun\n",
    "│       │   ├── example_demog_two_epoch_runXX.SD.txt\n",
    "│       │   ├── example_demog_two_epoch_runXX.info.txt\n",
    "│       │   ├── example_demog_two_epoch_runXX.png\n",
    "│       │   ├── example_demog_two_epoch_runXX.txt\n",
    "│       │   ├── example_demog_two_epoch_runXX_folded.expSFS\n",
    "│       │   └── example_demog_two_epoch_runXX_unfolded.expSFS\n",
    "│       ├── detail_5runs\n",
    "│       │   └── ...\n",
    "│       ├── example_demog_two_epoch_SFS.pdf\n",
    "│       └── example_demog_two_epoch_summary.txt\n",
    "└── logs\n",
    "    └── demography.log\n",
    "```\n",
    "\n",
    "* `example_demog_two_epoch_SFS.pdf`: The SFS fit for the best model\n",
    "* `example_demog_two_epoch_summary.txt`: Tabulated data for all the runs\n",
    "* `bestrun/`: the folder with all the information for the best run (where `XX` represents the run number e.g. `run03`)\n",
    "    * `example_demog_two_epoch_runXX.info.txt`: Tabulated data for the best run with uncertainty estimate appended (FIM + convergence)\n",
    "    * `example_demog_two_epoch_runXX.png`: `dadi`'s original diagnostic plot for examining residuals\n",
    "    * `example_demog_two_epoch_runXX_folded.expSFS`: folded expected SFS from this run"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## step2: `DFE1D_refspectra` module for building a reference spectra cache"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "WORKSCRIPT='../workflow/DFE/DFE1D_refspectra.py'\n",
    "\n",
    "# show help\n",
    "python $WORKSCRIPT -h\n",
    "\n",
    "# generate reference spectra cache\n",
    "python $WORKSCRIPT \\\n",
    "'two_epoch' '2.33,0.43' '100' './output/dfe/refspectra/example' \\\n",
    "&> './output/logs/dfe_refspectra.log'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Command line explanation\n",
    "\n",
    "Positional arguments:\n",
    "|Position | Value | Explanation |\n",
    "|---------|-------|-------------|\n",
    "|1 | 'two_epoch' | Demographic model to use |\n",
    "|2 | '2.33,0.43' | Demographic parameters (nu,T) for the two-epoch model |\n",
    "|3 | '100' | Number of samples in the SFS to generate |\n",
    "|4 | './output/dfe/refspectra/example' | Path/NamePrefix to the output file |\n",
    "\n",
    "Additional:\n",
    "- Redirects both stdout and stderr to './output/logs/dfe_refspectra.log'\n",
    "\n",
    "### Log outputs\n",
    "\n",
    "In the output log, it provides:\n",
    "\n",
    "1. Input parameter parsed\n",
    "2. The population-scaled selection coefficient values used for building each reference spectra\n",
    "3. Summary of the reference spectra cache\n",
    "\n",
    "### File outputs\n",
    "\n",
    "In the output folders, the files are organized as following:\n",
    "\n",
    "```\n",
    ".\n",
    "├── dfe\n",
    "│   └── refspectra\n",
    "│       ├── example_DFESpectrum.bpkl\n",
    "│       └── example_DFESpectrum_QC.pdf\n",
    "└── logs\n",
    "    └── dfe_refspectra.log\n",
    "```\n",
    "\n",
    "* `example_DFESpectrum.bpkl`: The reference spectra cache with given demographic parameters\n",
    "* `example_DFESpectrum_QC.pdf`: QC plots showing the expected SFS under various different selection coeffcients\n",
    "\n",
    "### Notes\n",
    "\n",
    "This step is time-consuming (~40 mins on a local computer for the example above). Using a cluster is recommended for efficiency.\n",
    "\n",
    "By default, both negative and positive selection coefficients are computed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## step3: `DFE1D_inferenceFIM` module for DFE inference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "WORKSCRIPT='../workflow/DFE/DFE1D_inferenceFIM.py'\n",
    "\n",
    "# show help\n",
    "python $WORKSCRIPT -h\n",
    "\n",
    "# perform DFE inference assuming a gamma-distributed DFE function\n",
    "python $WORKSCRIPT \\\n",
    "--pop 'example' --mu '2.50E-08' --Lcds '19089129' --NS_S_scaling '2.31' --Nrun 5 \\\n",
    "'MIS.sfs' './output/dfe/refspectra/example_DFESpectrum.bpkl' 'gamma' '4062' './output/dfe/gamma' &> './output/logs/dfe_gamma.log'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Command line explanation\n",
    "\n",
    "Positional arguments:\n",
    "|Position | Value | Explanation |\n",
    "|---------|-------|-------------|\n",
    "|1 | 'MIS.sfs' | Path to folded nonsynonymous SFS file |\n",
    "|2 | './output/dfe/refspectra/example_DFESpectrum.bpkl' | Path to reference DFE spectra |\n",
    "|3 | 'gamma' | DFE functional form to use |\n",
    "|4 | '4062' | Theta of synonymous regions |\n",
    "|5 | './output/dfe/gamma' | Output directory path |\n",
    "\n",
    "Options:\n",
    "|Argument | Value | Explanation |\n",
    "|---------|-------|-------------|\n",
    "|--pop | 'example' | Population identifier |\n",
    "|--mu | '2.50E-08' | Exon mutation rate (mutations/bp/generation) |\n",
    "|--Lcds | '19089129' | Number of called CDS sites used in SFS |\n",
    "|--NS_S_scaling | '2.31' | Scaling factor for Lsyn and Lnonsyn length |\n",
    "|--Nrun | 5 | Number of replicated runs |\n",
    "\n",
    "Additional:\n",
    "- Redirects both stdout and stderr to './output/logs/dfe_gamma.log'\n",
    "\n",
    "### Log outputs\n",
    "\n",
    "This script gives (hopefully) most of the information you need for DFE inference: \n",
    "\n",
    "In the output log, it provides: \n",
    "\n",
    "1. Input parameters parsed\n",
    "2. Within each replicated inference, every iteration's log-likelihood and DFE parameter values\n",
    "3. Top three replicates with the highest log-likelihoods\n",
    "4. Convergence of the best inference with the highest log-likelihoods\n",
    "5. Fisher's Information Matrix (FIM) based standard deviation estimates\n",
    "\n",
    "### File outputs\n",
    "\n",
    "In the output folders, the files are organized as following: \n",
    "\n",
    "```\n",
    ".\n",
    "├── dfe\n",
    "│   └── gamma\n",
    "│       ├── bestrun\n",
    "│       │   ├── example_DFE_gamma_runXX.SD.txt\n",
    "│       │   ├── example_DFE_gamma_runXX.info.txt\n",
    "│       │   ├── example_DFE_gamma_runXX.png\n",
    "│       │   ├── example_DFE_gamma_runXX.txt\n",
    "│       │   ├── example_DFE_gamma_runXX_folded.expSFS\n",
    "│       │   └── example_DFE_gamma_runXX_unfolded.expSFS\n",
    "│       ├── detail_5runs\n",
    "│       │   └── ...\n",
    "│       ├── example_DFE_gamma_PDF.pdf\n",
    "│       ├── example_DFE_gamma_SFS.pdf\n",
    "│       └── example_DFE_gamma_summary.txt\n",
    "└── logs\n",
    "    └── dfe_gamma.log\n",
    "```\n",
    "\n",
    "* `example_DFE_gamma_SFS.pdf`: The SFS fit for the best model\n",
    "* `example_DFE_gamma_PDF.pdf`: The probability density function (PDF) for the best DFE inference\n",
    "* `example_DFE_gamma_summary.txt`: Tabulated data for all the runs\n",
    "* `bestrun/`: the folder with all the information for the best run (where `XX` represents the run number e.g. `run03`)\n",
    "    * `example_DFE_gamma_runXX.info.txt`: Tabulated data for the best run with uncertainty estimate appended (FIM + convergence)\n",
    "    * `example_DFE_gamma_runXX.png`: `dadi`'s original diagnostic plot for examining residuals\n",
    "    * `example_DFE_gamma_runXX_folded.expSFS`: folded expected SFS from this run\n",
    "\n",
    "### Notes\n",
    "\n",
    "Currently, this pipeline supports the following DFE functions: \n",
    "\n",
    "1. gamma: Gamma distribution\n",
    "2. neugamma: Gamma distribution + neutral point mass \n",
    "3. gammalet: Gamma distribution + lethal point mass \n",
    "5. lognormal: Log-normal distribution\n",
    "6. lourenco_eq: Fisher's Geometric Model derived DFE distribution (eq. 15 in [Lourenço et al. 2011](https://onlinelibrary.wiley.com/doi/abs/10.1111/j.1558-5646.2011.01237.x))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## step4: `DFE1D_gridsearch` module for DFE grid-search and comparisons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "WORKSCRIPT='../workflow/DFE/DFE1D_gridsearch.py'\n",
    "\n",
    "# show help\n",
    "python $WORKSCRIPT -h\n",
    "\n",
    "python $WORKSCRIPT \\\n",
    "--max_bound \"0.5,0.5\" --min_bound \"0.1,1E-4\" --dfe_scaling --Nanc 7043 --Npts 10 \\\n",
    "'MIS.sfs' './output/dfe/refspectra/example_DFESpectrum.bpkl' 'gamma' '9383' './output/dfe/gridsearch/example_grids' &> './output/logs/dfe_gridsearch.log'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Command line explanation\n",
    "\n",
    "Positional arguments:\n",
    "|Position | Value | Explanation |\n",
    "|---------|-------|-------------|\n",
    "|1 | 'MIS.sfs' | Path to folded nonsynonymous SFS file |\n",
    "|2 | './output/dfe/refspectra/example_DFESpectrum.bpkl' | Path to reference DFE spectra |\n",
    "|3 | 'gamma' | DFE functional form to use |\n",
    "|4 | '9383' | Theta of nonsynonymous regions |\n",
    "|5 | './output/dfe/gridsearch/example_grids' | Path/NamePrefix to the output file |\n",
    "\n",
    "Options:\n",
    "|Argument | Value | Explanation |\n",
    "|---------|-------|-------------|\n",
    "|--max_bound | \"0.5,0.5\" | Maximum values range for DFE parameters |\n",
    "|--min_bound | \"0.1,1E-4\" | Minimum values range for DFE parameters |\n",
    "|--dfe_scaling | enabled | Use `2NeS` or `S` distribution (flag present, so unscale the population-scaled DFE `2NeS` to individual-level `S`) |\n",
    "|--Nanc | 7043 | Ancestral population size for DFE scaling |\n",
    "|--Npts | 10 | Number of grid points per parameter |\n",
    "\n",
    "Additional:\n",
    "- Redirects both stdout and stderr to './output/logs/dfe_gridsearch.log'\n",
    "\n",
    "This command performs a grid search for DFE parameters using the specified reference spectra and SFS file. It uses a gamma distribution as the DFE functional form, applies DFE unscaling, and outputs the likelihood surface to the specified directory.\n",
    "\n",
    "### Log outputs\n",
    "\n",
    "In the output log, it provides: \n",
    "\n",
    "1. Input parameters parsed\n",
    "2. Highest log-likelihoods during grid search\n",
    "3. Final execution status \n",
    "\n",
    "### File outputs\n",
    "\n",
    "In the output folders, the files are organized as following: \n",
    "\n",
    "```\n",
    ".\n",
    "├── dfe\n",
    "│   └── gridsearch\n",
    "│       ├── example_grids.npy\n",
    "│       ├── example_grids.pdf\n",
    "│       └── example_grids.txt\n",
    "└── logs\n",
    "    └── dfe_gridsearch.log\n",
    "```\n",
    "\n",
    "* `example_grids.npy`: The DFE grid-search output in numpy formats\n",
    "* `example_grids.txt`: The DFE grid-search output in text formats\n",
    "* `example_grids.pdf`: Likelihood surface and MLE estimates derived from the grid-search"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# About\n",
    "\n",
    "## Disclaimer\n",
    "\n",
    "varDFE and this tutorial are provided \"as is\" without any warranties or representations of any kind, express or implied. I make no guarantees or warranties regarding the accuracy, reliability, completeness, suitability, or timeliness of the software.\n",
    "\n",
    "## Citation\n",
    "\n",
    "When using varDFE, please also cite the dadi and fitdadi packages:\n",
    "\n",
    "RN Gutenkunst, RD Hernandez, SH Williamson, CD Bustamante \"Inferring the joint demographic history of multiple populations from multidimensional SNP data\" PLoS Genetics 5:e1000695 (2009).\n",
    "\n",
    "BY Kim, CD Huber, KE Lohmueller \"Inference of the Distribution of Selection Coefficients for New Nonsynonymous Mutations Using Large Samples\" Genetics 206:345 (2017)."
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
